### PR TITLE
[MRG] estimated step size for SAG* objects

### PR DESCRIPTION
**Pull request by fabianp**
This estimates the step size from the data instead of relying on the user having to specify a step size. The result is that it converges for many cases for which the default step size will not converge, such as [multiplying the 20news dataset by 3](https://gist.github.com/fabianp/1c80fc522617f3d08f6d).

Some remarks:

* The modified Huber loss is not Lipschitz-smooth so the step size is not well defined (AFAIK the theory does not guarantee convergence in this case). The step size that we propose in this case is given taking a bound on the second derivative ignoring the discontinuities, which seems to work fine in practice.

* The hard work is done by @zermelozf, I just did cosmetic changes.

### CODE REVIEW

File: lightning/impl/sag.py [hunk 1]
```diff
@@ -1,8 +1,20 @@
...
+try:
+    from sklearn.linear_model.sag_fast import get_max_squared_sum
+except ImportError:
+    # compatibility for older scikit-learn
+    # this is orders of magnitude slower than the one in scikit-learn,
+    # but does not need to convert X to a dense matrix
+    def get_max_squared_sum(X):
+        n = X.shape[0]
+        return np.max(X.dot(X.T)[range(n), range(n)])
+
```
Comment (mblondel) [inline lightning/impl/sag.py]: I would use `np.sum(X ** 2, axis=1).max()` in the dense case and `csr_row_norms(X).max()` in the CSR case (from sklearn.utils.sparsefuncs).

**Comment by mblondel** [comment]
Very nice! Thanks!

File: lightning/impl/sag.py (hunk 1 above)
Comment (fabianp) [inline lightning/impl/sag.py]: I think that function is available in [scikit-learn 0.14](https://github.com/scikit-learn/scikit-learn/tree/0.14.X/sklearn/utils), it looks like it was added in 0.15

Comment (fabianp) [inline lightning/impl/sag.py]: I used your formula for the dense case at least.

Comment (mblondel) [inline lightning/impl/sag.py]: We can bump the dependency to 0.15 (released in July 2014).

**Comment by fabianp** [comment]
I've changed the backport of get_max_squared_sum to use csr_row_norms as suggested by @mblondel . I've also updated the README to reflect this (the import is anyway lazy so lightning should at least import in older versions of sklearn).

**Comment by mblondel** [comment]
Merging, thank you very much!
We can add this for SVRG too but this can be done in another PR.
